File: DataStructure/Search&Sort/heap_sort.py
# 堆排序实现 (大顶堆)

def heap_sort(arr):
    sz = len(arr)

    # 构建大顶堆
    for i in range(sz // 2 - 1, -1, -1): # 从第一个非叶子节点开始，从上到下调整
        adjust_heap(arr, i , sz) # 自上而下调整

    # 迭代调整堆顺序
    for j in range(sz - 1, 0, -1):
        swap(arr, 0, j) # 将堆顶移到数组末尾，即将最大值移到数组末尾
        adjust_heap(arr, 0, j) # 对堆顶元素进行调整，恢复二叉堆


def adjust_heap(arr, i, sz):
    tmp = arr[i]
    k = i * 2 + 1
    while k < sz:
        if k + 1 < sz and arr[k] < arr[k + 1]:
            k += 1
        if arr[i] < arr[k]: # 若父节点 < 子节点，则下沉父节点，上浮子节点
            arr[i] = arr[k]
            i = k
        else: 
            break

        k = k * 2 + 1

    arr[i] = tmp

def build_heap(arr):
    sz = len(arr)
    for i in range(sz // 2 - 1, -1, -1):
        adjust_heap(arr, i, sz)

def del_max(heap):
    sz = len(heap)
    swap(arr, 0, sz - 1) # 将最大节点与最后一个节点交换
    heap.pop() # 删除最大节点
    adjust_heap(arr, 0, sz - 1) # 调整堆，恢复为大顶堆


def swap(arr, i, j):
    arr[0], arr[j] = arr[j], arr[0]

if __name__ == "__main__":
    arr = [3,6,-1,4,5]
    heap_sort(arr)
    print(arr)

File: DataStructure/Search&Sort/heap_sort.cpp
//
// Created by 薛智钧 on 2020/4/5.
//
#include <cstdio>

static void adjustHeap(int arr[], int i, int len){
    int tmp = arr[i]; // 缓存当前元素
    for (int k = i * 2 + 1; k < len; k = k * 2 + 1){ // 从元素i的左孩子开始
        if (k + 1 < len && arr[k] < arr[k + 1]) ++k; //若左孩子节点小于右孩子，则将k指向右孩子

        if (tmp < arr[k]){ // 若子节点值大于父节点，将子节点值赋值给父节点，不用交换
            arr[i] = arr[k];
            i = k;
        }else{
            break;
        }
    }
    arr[i] = tmp;   // tmp放到最终位置
}

static void swap(int arr[], int a, int b){
    int tmp = arr[a];
    arr[a] = arr[b];
    arr[b] = tmp;
}

void heapSort(int arr[], int len){
    // 1. 构建大顶堆
    for (int i = len / 2 - 1; i >= 0; --i){
        // 从第一个非叶子节点开始，从上到下调整
        adjustHeap(arr, i, len);
    }

    // 迭代调整堆结构
    for (int j = len - 1; j > 0; --j){
        swap(arr, 0, j); // 堆顶和堆尾元素互换
        adjustHeap(arr, 0, j); // 对当前堆顶元素调整
    }
}

static void printArray(int arr[], int len){
    int i = 0;
    for(; i < len - 1; ++i)
        printf("%d, ", arr[i]);
    printf("%d", arr[i]);
    printf("\n");
}

int main(){
    int arr[] = {3,9,7,5,6,8,8};
    int len = sizeof(arr) / sizeof(arr[0]);

    printf("The given array is: \n");
    printArray(arr, len);
    heapSort(arr, len);
    printf("The sorted result is: \n");
    printArray(arr, len);
    return 0;
}


